File: code/testml/testml.rb
class TestML
  VERSION = '0.0.2'

  attr_accessor :runtime
  attr_accessor :compiler
  attr_accessor :bridge
  attr_accessor :library
  attr_accessor :testml

  def initialize attributes={}
    attributes.each { |k,v| self.send "#{k}=", v }
    yield self if block_given?
  end

  def run(*args)
    set_default_classes
    @runtime.new(
      compiler: @compiler,
      bridge: @bridge,
      library: @library,
      testml: @testml,
    ).run(*args)
  end

  def set_default_classes
    if not @runtime
      require 'testml/runtime/unit'
      @runtime = TestML::Runtime::Unit
    end
    if not @compiler
      require 'testml/compiler/pegex'
      @compiler = TestML::Compiler::Pegex
    end
    if not @bridge
      require 'testml/bridge'
      @bridge = TestML::Bridge
    end
    if not @library
      require 'testml/library/standard'
      require 'testml/library/debug'
      @library = [
        TestML::Library::Standard,
        TestML::Library::Debug,
      ]
    end
  end
end

File: code/testml/testml.py
__version__ = '0.0.2'

class TestML:
    def __init__(self, **args)
        self.testml = args.get('testml')
        self.runtime = args.get('runtime')
        self.bridge = args.get('bridge')
        self.library = args.get('library')
        self.compiler = args.get('compiler')

    def run(self, *args):
        self.set_default_classes()
        self.runtime(
            compiler=self.compiler,
            bridge=self.bridge,
            library=self.library,
            testml=self.testml,
        ).run(*args)

    def set_default_classes(self):
        if not self.runtime:
            from testml.runtime.unit import Unit
            self.runtime = Unit
        if not self.compiler:
            from testml.compiler.pegex import Pegex
            self.compiler = Pegex
        if not self.bridge:
            from testml.bridge import Bridge
            self.bridge = Bridge
        if not self.library:
            from testml.library.standard import Standard
            from testml.library.debug import Debug
            self.library = [
                Standard,
                Debug,
            ]
